Rewrite rust lib

havardh opened the pull request.

havardh committed:
Draft to rewrite of sections describing REL, RCL and RSL

havardh committed:
Merge branch 'master' into rewrite-rust-lib

havardh committed:
Removed typeo

havardh committed:
Fixed text

havardh committed:
Merge branch 'master' into rewrite-rust-lib

# Conflicts:
#	background/rust.tex
#	implementation/building.tex

havardh committed:
Removed String and Vector section

sondrele commented on background/rust.tex:
```diff
@@ -11,18 +11,33 @@ \section{The Rust Programming Language} % (fold)
 \label{ssub:rust:features}
 
-This section starts by describing {\rust}'s {\core} library and its type-system, before moving on to presenting some of the language features that make {\rust} so different from C.
+This section starts by describing {\rust}'s standard library and its type-system, before moving on to presenting some of the language features that makes {\rust} so different from {\C}.
```
mulig det skal være med `make rust so different from c`, ikke `makes?`

sondrele commented on background/rust.tex:
```diff
@@ -11,18 +11,33 @@ \section{The Rust Programming Language} % (fold)
-This library is not directly accessible to the programmer, instead its stable parts are reexported through {\rust}'s standard library called {\std}.
-In this way, {\core} is not interfaced with directly in normal {\rust} programs, so it can be modified while {\std}'s interface stays the same.
-The standard library also reexports a handful of other libraries and does the task of implementing platform-dependent features for all of its target \glspl{os}.
+The \gls{rsl} is included into all {\rust} programs and strives to be an abstraction layer to write portable applications.
```
Vet ikke helt med `... and strives to be an abstraction layer to write...`.
Er std ment som et faktisk "lag med abstraksjoner", eller bare som en "paraply" for Rust's forskjellige bibliotker?

havardh replied:
Vel, std er to ting (som irriterer meg litt) 1) en facade for alt som ligger under `collections` osv og 2) litt ekstra implementatsjon av `fs` og ? `os` og sånt. Men synes det er funker med denne høynivås beskrivelsen. Std prøver å la deg skrive et Rust program som kan kjøre på flere platformer, i.e. et abstraksjons lag.

sondrele commented on background/rust.tex:
```diff
@@ -11,18 +11,33 @@ \section{The Rust Programming Language} % (fold)
-In this way, {\core} is not interfaced with directly in normal {\rust} programs, so it can be modified while {\std}'s interface stays the same.
-The standard library also reexports a handful of other libraries and does the task of implementing platform-dependent features for all of its target \glspl{os}.
+The \gls{rsl} is included into all {\rust} programs and strives to be an abstraction layer to write portable applications.
+The library is implemented as a facade, meaning that the crate {\std} provides a thin wrapper which reexports a stable interface for crates like \lib{alloc} (for memory allocation), \lib{collections} (general purpose data structures) and {\core} (described in the next section).
```
Kanskje litt uheldig å bruke ordet `crate` før `Organization`-seksjonen?

sondrele commented on background/rust.tex:
```diff
@@ -11,18 +11,33 @@ \section{The Rust Programming Language} % (fold)
+
+The distinction between \gls{rcl} and the {\core} crate is not that clear as the {\std} and \gls{rsl} as \gls{rcl} does not export any functionallity which is not defined in {\core}.
+
+\autoref{tab:rust:datatypes} lists the primitive datatypes available in {\rust}, \gls{rcl} defines many functions and abstractions that are used to manage and manipulate these datatypes.
```
Ok, den lister de forskjellige typene, men hvorfor er de gjengitt i denne tabellen? 

Jeg synes setningen under forklarer hvorfor tabellen er der.
Mulig vi ikke trenger den i det hele tatt?

sondrele commented on background/rust.tex:
```diff
@@ -11,18 +11,33 @@ \section{The Rust Programming Language} % (fold)
+This gives the developer the ability to modify the library while the \gls{rsl} interface stays the same.
+%The standard library also reexports a handful of other libraries and does the task of implementing platform-dependent features for all of its target \glspl{os}.
+
+The distinction between \gls{rcl} and the {\core} crate is not that clear as the {\std} and \gls{rsl} as \gls{rcl} does not export any functionallity which is not defined in {\core}.
```
Denne setningen er kanskje litt rar, kan kanske endres til dette: `the distinction between RCL and the core crate is not as clear as the distinction between std and RSL, as RCL does not export any functionality which is not defined in core.`

Også, når det nevnes at "this distinction is not as clear" så burde det kanskje nevnes _hva_ forskjellen er?

Tror du den er nødvendig i det hele tatt?

sondrele commented on implementation/building.tex:
```diff
@@ -95,7 +95,7 @@ \subsection{Transitioning to Cargo}
 This problem was solved by implementing a new Cargo build-dependency, called \texttt{rust-src} \cite{github:rust_src}, whose purpose is to download the entire {\rust} source code that is compliant with the instance of {\rustc} that is \emph{currently} compiling the library.
 By making it a task for each build to fetch its own source code, we were guaranteed that the dependencies we used for the project would always compile, independent of the current instance of {\rustc} that was installed on the system.
-The crates that we have fetched from {\rust}'s standard library that make up what we call \gls{rel} are already described in \autoref{sec:rust_embedded_library}, but they are also shown in \autoref{tab:compiled_libraries} for the sake of completeness.
+The crates that we have fetched from {\rust}'s standard library that make up what we call \gls{rel} is already described in \autoref{sec:rel}, but they are also shown in \autoref{tab:compiled_libraries} for the sake of completeness.
```
Jeg tror `are` er riktig istedet for `is`?

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
 
-When we consider {\std}'s dependencies, shown in \autoref{fig:rust:librust}, we see that the crate is not suitable for a bare-metal application because it depends on an \gls{os} to be present.
-The dependency on either a Unix or a Windows \gls{os} excludes {\std} and all of its dependents from being used directly with the EFM32.
+As mentioned in \autoref{sec:rsl}, \gls{rsl} is built with the assumtion that the application is running on an OS.
```
assumtion -> assumption

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
-The dependency on either a Unix or a Windows \gls{os} excludes {\std} and all of its dependents from being used directly with the EFM32.
+As mentioned in \autoref{sec:rsl}, \gls{rsl} is built with the assumtion that the application is running on an OS.
 This section presents some of the libraries provided by {\rust} that can be used in a bare-metal system, and how and why they are composed into what we refer to as the \gls{rel}.
+\gls{rel} is unlike \gls{rsl} not built as a facade, in fact it is not implemented at all.
```
Komma: `REL is, unlike RSL, not ...`

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
+As mentioned in \autoref{sec:rsl}, \gls{rsl} is built with the assumtion that the application is running on an OS.
 This section presents some of the libraries provided by {\rust} that can be used in a bare-metal system, and how and why they are composed into what we refer to as the \gls{rel}.
+\gls{rel} is unlike \gls{rsl} not built as a facade, in fact it is not implemented at all.
+It it just a way to provide a well defined definition of the {\rust} library in an embedded system.
```
It it just -> It is just?

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
     \includegraphics[scale=0.3]{figures/background/rust/embedded-rust-lib.png}
   \end{center}
-  \caption{\rust Embedded Library}
+  \caption{{\rust} Embedded Library}
```
Denne figuren blir ikke referert til...

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
-
-The {\rust} collection library provides the data structures shown in \autoref{tab:rust:collections}.
+As described in \autoref{sec:rcl} the \gls{rcl} defines the \emph{core} functionality of the {\rust} language.
+These includes the primitive types listed in \autoref{tab:rust:datatypes} in addition to some key datastructures as \code{Option} and \code{Result} as described in \autoref{ssub:rust:features}.
```
Som Toni har påpekt et par ganger, så er det kanskje litt uheldig å refere til tabeller og figurer som er så langt unna? Trenger typene og Option egentlig å gjentas her?

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
-    String \\
+  \begin{tabular}{l l}
+    \code{memcpy, memcmp, memset} & basic memory management \\
+    \code{rust\_begin\_unwind} & handling panicing \\
```
panicing -> panicking

sondrele replied:
Kanskje det passer med en setning her om hva en panic er her?

f.eks. `Handles panicking. A panic is Rust's way of unwinding the currently executing thread, evidently resulting in the thread being terminated` eller noe ligenende.

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
-We look at some of the data structures in this section.
-
-\subsubsection{Vector}
+The memory management functions given in \autoref{tab:core:definitions} are provided by \lib{newlib} and are exposed thought the \lib{startup} library described in \autoref{sec:startup}.
```
endre til: `... and are exposed through the ...`

også, denne fungerer ikke: `\autoref{sec:startup}`

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
   \end{tabular}
-  \caption{\rust Collection Library}
-  \label{tab:rust:collections}
+  \caption{}
```
Trenger caption

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
+Heap allocation is introduced in a library called \lib{alloc}.
+The library defines the owned pointer \code{Box} which is {\rust}'s main means of allocating memory on the heap.
+In addition the allocation library defines the types \code{Rc} and \code{Arc} which are {\rust}'s reference counted and atomically reference counted heap pointers.
+The allocation library is by default dependent on \lib{libc}, but by suppling the \flag{--cfg feature="external\_funcs"} this dependency can be broken.
```
Kanskje endre til: `on libc, but this dependency can be broken by supplying the --cfg ... flag to the compilation procedure.`

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
+  \begin{minted}{rust}
+rust_allocate(usize, usize) -> *mut u8;
+rust_deallocate(*mut u8, usize, usize);
+rust_reallocate(*mut u8, usize, usize, usize) -> *mut u8;
```
Kanskje legge på fn?

```
fn rust_allocate(usize, usize) -> *mut u8;
fn rust_deallocate(*mut u8, usize, usize);
fn rust_reallocate(*mut u8, usize, usize, usize) -> *mut u8;
```

sondrele commented on implementation/rust-embedded-library.tex:
```diff
@@ -1,65 +1,62 @@
-- Reference Counted GC
+The {\rust} collection library provides general purpose data structures.
+Of which the \code{Vector}, a growable heap allocated list, and the \code{String}, heap allocated mutable strings, are the most notable.
+As one would guess the collection library depends on the allocation library as it needs to allocate memory on the heap.
```
komma: `As one would guess, ...`

havardh committed:
Fixed typeo

havardh committed:
Removed distinction between std and RSL + core and RCL.
It done as the distinction is not relevant (we do not use RSL) and to remove the `crate` concept before it is introduced

havardh committed:
Added rational for primitive table in Rust Core description

havardh committed:
Fixed typeo's a figures

havardh committed:
Remove 'broken' figure.

havardh committed:
Added startup section

havardh committed:
Changed wording
